docs/tutorials/introduction.rst

```diff
@@ -0,0 +1,307 @@
+.. _tutorial-introduction:
+
+******************************************************
+Introduction to Python packaging with ``meson-python``
+******************************************************
+
+If you are new to Python packaging, don't worry!
+
+We will give you a quick introduction to what steps releasing a Python package
+consists of, and walk you through them to get started.
+
+
+Create a Meson_ project
+=======================
+
+To get started, we need a project to publish. As ``meson-python`` is built on
+top of Meson, we will create a really simple Meson project. You may already have
+a Meson project you wish to publish, in that case, you can simply skip this
+step.
+
+
+The module
+----------
+
+First, we create a simple Python module. We will go for a native module, as
+that's really where ``meson-python`` shines against other Python build backends.
+
+
+.. code-block:: c
+   :caption: our_first_module.c
+
+   #include <Python.h>
+
+   static PyObject* foo(PyObject* self)
+   {
+       return PyUnicode_FromString("bar");
+   }
+
+   static PyMethodDef methods[] = {
+       {"foo", (PyCFunction)foo, METH_NOARGS, NULL},
+       {NULL, NULL, 0, NULL},
+   };
+
+   static struct PyModuleDef module = {
+       PyModuleDef_HEAD_INIT,
+       "our_first_module",
+       NULL,
+       -1,
+       methods,
+   };
+
+   PyMODINIT_FUNC PyInit_our_first_module(void)
+   {
+       return PyModule_Create(&module);
+   }
+
+
+Here, we have a create a small module named ``our_first_module``, which has a
+function ``foo`` that simply returns ``"bar"``.
+
+
+.. admonition:: Using the C API
+   :class: seealso
+
+   If you need help writing a native module using Python's C API, we recommend
+   checking out the following resources.
+
+   `The official Python C API reference manual <https://docs.python.org/3/c-api/index.html>`_
+   `RealPython's "Building a Python C Extension Module" introductory tutorial <https://realpython.com/build-python-c-extension-module/>`_
+   `pysheeet's "C Extensions" page <https://www.pythonsheets.com/notes/python-c-extensions.html>`_
+   `pysheeet-kr's "Python C API cheatsheet" page <https://pysheeet-kr.readthedocs.io/ko/latest/notes/python-capi.html>`_
+
+
+The Meson build description
+---------------------------
+
+Now, we need to create the Meson build description file. This tells Meson what
+we want it to build, and how to do it.
+
+
+.. code-block:: meson
+   :caption: meson.build
+
+   project('purelib-and-platlib', 'c')
+
+   py_mod = import('python')
+   py = py_mod.find_installation()
```

Comment:
From https://mesonbuild.com/Python-module.html#find_installation

>  `pure`: On some platforms, architecture independent files are expected to be placed in a separate directory. However, if the python sources should be installed alongside an extension module built with this module, this keyword argument can be used to override the default behavior of `.install_sources()`. since 0.64.0 

Hummm... I do not follow. AFAICT this will just force sources to be placed in platlib.

Reply:
Which is what you want, since you have C code in your package. This is a safety measure; otherwise you have to plaster `pure: false` everywhere you use `py.install_sources`, and if you forget it just once you end up with a broken package. That's why it was added: https://mesonbuild.com/Release-notes-for-0-64-0.html#pythonfind_installation-now-accepts-pure-argument

Reply:
Why do you end up with a broken package? Is there an assumption somewhere that all the code in a package needs to be installed either in `{platlib}` or in `{purelib}`? meson-python does not do this assumption, and actually has code to handle this case. If this is the case, it would be nice to detect the circumstance and raise an error in meson-python. However, I don't see why this is the case. Also, in most platforms, `{platlib}` and `{purelib}` are the same path.

Reply:
> Is there an assumption somewhere that all the code in a package needs to be installed either in `{platlib}` or in `{purelib}`?

Yes of course. Say you have a `foo.py` file and `bar.so` extension module, and you do a relative import `from .bar` in `foo.py`, that only works if the two files are in the same directory. Which is the kind of thing you only see is broken when some niche Linux distro doesn't have purelib and platlib pointing to the same path (not Ubuntu, which is typically the one thing people test in their CI). This whole purelib/platlib concept is just awful.

> meson-python does not do this assumption, and actually has code to handle this case. If this is the case, it would be nice to detect the circumstance and raise an error in meson-python.

Not every build goes only through `meson-python`. If it's broken for plain `meson setup/install`, that's still bad. Raising an error SGTM.

Reply:
Ah, that's obvious, but I can have a wheel that installs multiple top-level modules. And those can be installed some in `{platlib}` and some in `{purelib}`. Saying that if you don't pass `pure: false` you get a broken wheel is a simplification that I don't think we should make.

Reply:
Multiple top-level modules is a corner case that we're not dealing with here. I'm not asking to say _" if you don't pass `pure: false` you get a broken wheel "_. I am asking to please use the robust `pure: false` as the default everywhere, instead of the very fragile state of not doing that. Which I'm quite sure is the right thing to do (for more details, please see https://github.com/mesonbuild/meson/issues/10523).

Reply:
Setting this on the installation object will flip the default value on the install_sources() and get_install_dir() functions. This is the correct, sensible, sane, and *only* reasonable thing to do if you have a top-level module with, nested inside it, a mixture of pure python and compiled extensions: it means that by default, you install pure python together with the compiled extension.

It doesn't prevent you from passing `pure: true` to individual calls where you do intend to install some freestanding top-level pure-python modules to purelib anyways. And I agree, that's a valid use case -- although I expect it's also *incredibly* rare.

Failing to place pure python files in the platlib when they are nested inside a package together with platlib extensions, means that you end up with one package being a namespace package and the other not, with corresponding... abuses... of the importlib search path. You are not going to end up happy if that happens, because at least one user is going to be on a distro which splits the two, and then disaster is going to happen and it's not going to be pretty.

We should absolutely document the default recommended approach, as the one that doesn't break people. :)

Reply:
I think it's reasonable to set that, but IMO it would be best to not set it, and add an admonition below pointing to a page that explains this in-depth.

> Failing to place pure python files in the platlib when they are nested inside a package together with platlib extensions, means that you end up with one package being a namespace package and the other not, with corresponding... abuses... of the importlib search path. You are not going to end up happy if that happens, because at least one user is going to be on a distro which splits the two, and then disaster is going to happen and it's not going to be pretty.

We should try to detect these cases and warn users.

Reply:
I added this

```rst
.. admonition:: Mixing pure and native code
   :class: caution

   If you are mixing pure (``.py``) and native modules in your project, please
   check out :ref:`reference-quirks-mixing-purelib-and-platlib`, which explains
   the common foot-guns when doing so, and how to make sure you install your
   modules correctly.
```

Reply:
@FFY00, I disagree - this is not a choice between two equivalent alternatives. `pure: false` is so far better, it must be used. You have no idea how much time this has already cost me in subtle bugs, and how tricky it is to debug. Please just accept the comment. Keeping a bad default and then adding cognitive load for the user with a note explaining why they should not follow your example code because of footguns is not useful.

Reply:
I changed it.

IMO this is the kind of issue we can, and should be trying to detect.